apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/ReliableStreamReader.cs

```diff
@@ -0,0 +1,258 @@
+                    SpannerSettings.GetDefaultRetryBackoff(),
+                    SpannerSettings.GetDefaultTimeoutBackoff(),
+                    Expiration.FromTimeout(TimeSpan.FromMilliseconds(600000)),
+                    SpannerSettings.IdempotentRetryFilter
+                )));
+            _request.SessionAsSessionName = _session.SessionName;
+        }
+
+        private async Task<Metadata> ConnectAsync()
+        {
+            Logger.LogPerformanceCounterFn("StreamReader.ConnectCount", x => x + 1);
+            if (_resumeToken != null)
+            {
+                Logger.Debug(() => $"Resuming at location:{_resumeToken}");
+                _request.ResumeToken = _resumeToken;
+            }
+            _currentCall = _spannerClient.ExecuteSqlStream(_request);
+            return await _currentCall.ResponseHeadersAsync.ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Connects or reconnects to Spanner, fast forwarding to where we left off based on
+        /// our _resumeToken and _resumeSkipCount.
+        /// </summary>
+        /// <returns></returns>
+        private async Task<bool> ReliableConnect(CancellationToken cancellationToken)
+        {
+            if (_currentCall == null)
+            {
+//                Func<Task<Metadata>> connectFn = ConnectAsync;
+//                var responseHeaders = await connectFn.CallWithRetry(_callTiming, _clock, _scheduler);
+                await ConnectAsync().ConfigureAwait(false);
+
+                Debug.Assert(_currentCall != null, "_currentCall != null");
+                cancellationToken.ThrowIfCancellationRequested();
+
+                for (int i = 0; i <= _resumeSkipCount; i++)
+                {
+                    //This calls a simple movenext on purpose.  If we get an error here, we'll fail out.
+                    _isReading = await _currentCall.ResponseStream.MoveNext(cancellationToken);
+                    if (!_isReading || _currentCall.ResponseStream.Current == null)
+                    {
+                        return false;
+                    }
+                    if (_metadata == null)
+                    {
+                        _metadata = _currentCall.ResponseStream.Current.Metadata;
+                    }
+                }
+                RecordResumeToken();
+            }
+            return _isReading;
+        }
+
+        private async Task<bool> ReliableMoveNext(CancellationToken cancellationToken)
+        {
+            try
+            {
+                Logger.LogPerformanceCounterFn("StreamReader.MoveNextCount", x => x + 1);
+                //we increment our skip count before calling MoveNext so that a reconnect operation
+                //will fast forward to the proper place.
+                _resumeSkipCount++;
+                _isReading = await _currentCall.ResponseStream.MoveNext(cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception e) 
```

Comment:
But how about IOException or having the broken connection manifest in a different way?  I understand the objection, but in practical terms, if there is a persistent issue, it will manifest again immediately when a reconnect is attempted and then we'll bail.
Feels like we are more likely to have someone alter exceptions underneath us and break the reliable-ness than anything bad happening because of catching a general exception.  We do, after all log it anyway.

If only c# had checked exceptions like java, I'd feel differently.